homeassistant/components/sensoterra/coordinator.py

```diff
@@ -0,0 +1,136 @@
+                self.timestamp = probe.timestamp
+        elif type == "LASTSEEN":
+            self.value = probe.timestamp
+            self.timestamp = datetime.now(UTC)
+        elif type == "RSSI":
+            if probe.rssi is not None:
+                self.value = round(probe.rssi, 0)
+                self.timestamp = probe.timestamp
+        else:
+            raise UpdateFailed(f"Unknown sensor type {type}")
+        if self.timestamp is None or self.timestamp < expiration:
+            self.value = None
+
+
+class SensoterraCoordinator(DataUpdateCoordinator):
+    """Sensoterra coordinator."""
+
+    add_devices_callback: Callable[[dict[str, SensoterraSensor]], None] | None
+
+    def __init__(self, hass: HomeAssistant, api: CustomerApi) -> None:
+        """Initialize Sensoterra coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            # Name of the data. For logging purposes.
+            name="Sensoterra probe",
+            # Polling interval. Will only be polled if there are subscribers.
+            update_interval=timedelta(seconds=SCAN_INTERVAL),
+        )
+        self.api = api
+        self.add_devices_callback = None
+
+    async def _async_update_data(self) -> dict[str, SensoterraSensor]:
+        """Fetch data from Sensoterra Customer API endpoint."""
+        try:
+            current_sensors = set(self.async_contexts())
+            probes = await self.api.poll()
+        except ApiAuthError as err:
+            # Raising ConfigEntryAuthFailed will cancel future updates
+            # and start a config flow with SOURCE_REAUTH (async_step_reauth)
+            raise ConfigEntryAuthFailed(err) from err
+        except ApiTimeout as err:
+            raise UpdateFailed("Timeout communicating with Sensotera API") from err
+        else:
+            sensors = {}
+            expiration = datetime.now(UTC) - timedelta(seconds=SCAN_INTERVAL)
+            for probe in probes:
+                for sensor in probe.sensors():
+                    if sensor.type == "MOISTURE" and sensor.unit == "SI":
+                        sensor_id = f"{sensor.id}-SI"
+                        sensor_type = "SI"
+                    else:
+                        sensor_id = sensor.id
+                        sensor_type = sensor.type
+                    sensors[sensor_id] = SensoterraSensor(
+                        expiration, probe, sensor_type, sensor
+                    )
+                if probe.battery is not None:
+                    sensor_id = f"{probe.serial}-BATTERY"
+                    sensors[sensor_id] = SensoterraSensor(expiration, probe, "BATTERY")
+                if probe.rssi is not None:
+                    sensor_id = f"{probe.serial}-RSSI"
+                    sensors[sensor_id] = SensoterraSensor(expiration, probe, "RSSI")
+                sensor_id = f"{probe.serial}-LASTSEEN"
+                sensors[sensor_id] = SensoterraSensor(expiration, probe, "LASTSEEN")
```

Comment:
Together with the SemsoterraSensor object this feels like code that should be in the library rather than here

Reply:
I've simplified the code. The library uses a slightly different data model (a probe has a set of sensors, where HA uses a flat list of sensors instead), so some work needs to be done here.